(PDOC-28) Make puppet strings available as a gem

danzilio opened the pull request:
This is a reimplementation of #34 so as to be more maintainable :)

iankronquist commented:
@danzilio Thanks for the PR and sorry to leave the previous one open for so long. I'm going to test it out this morning.

danzilio commented:
no worries :)

iankronquist commented on puppet-strings.gemspec:
```diff
@@ -0,0 +1,18 @@
+puppet_metadata = JSON.load File.open(File.expand_path(File.join(__FILE__, '..', 'metadata.json'))).read
+
+Gem::Specification.new do |s|
+    s.name = 'puppet-strings'
```
I think this should be `'puppetlabs-strings'`. I could be mistaken though.

iankronquist replied:
Actually, maybe we should get this from the metadata as well?

iankronquist commented on puppet-strings.gemspec:
```diff
@@ -0,0 +1,18 @@
+    s.files = Dir['lib/**/*'].reject { |f| f if File.directory?(f) }
+
+    s.add_runtime_dependency 'puppet', '>= 3.7.0'
+    s.add_runtime_dependency 'yard'
```
We should pin ourselves to a yard version. Our incestuous relationship with the library means that changes to yard private apis could break certain monkeypatches in strings (gross). I'm going to go ahead and recommend we pin to '0.8.x', but I would like @HAIL9000's input when she gets back.

This should be documented elsewhere too.

danzilio commented:
@iankronquist I'll touch up the gemspec. I called it 'puppet-strings' because it seemed a little more friendly for mere mortals. I'm not opposed to calling the gem 'puppetlabs-strings'.

iankronquist commented on README.md:
```diff
@@ -164,6 +164,16 @@ Here are a few other good resources for getting started with documentation:
+
+To use the rake tasks, `require puppet-strings/rake_tasks` in your `Rakefile`:
+
+    require 'puppet-strings/rake_tasks'
```
nitpick: Can you mark this as ruby so we get fancy syntax highlighting? I should really do this elsewhere in the file too.
This applies to line 172 too.

iankronquist commented:
@danzilio Although I agree that name is more  friendly, I think consistency trumps friendliness in this case.

danzilio commented:
@iankronquist should be good now! let me know what you think!

iankronquist commented:
Okay, I here's how I'm testing this. Let me know if you're doing something different.

```
iankronquist@puppettop:(strings)(pr/54) → bundle exec gem build puppet-strings.gemspec 
WARNING:  no email specified
WARNING:  description and summary are identical
  Successfully built RubyGem
  Name: puppetlabs-strings
  Version: 0.2.0
  File: puppetlabs-strings-0.2.0.gem
iankronquist@puppettop:(strings)(pr/54) → bundle exec gem install puppetlabs-strings-0.2.0.gem 
Successfully installed puppetlabs-strings-0.2.0
1 gem installed
Installing ri documentation for puppetlabs-strings-0.2.0...
Building YARD (yri) index for puppetlabs-strings-0.2.0...
[error]: Unhandled exception in YARD::Handlers::Ruby::ClassHandler:
[error]:   in `lib/puppet_x/puppetlabs/strings/yard/handlers/base.rb`:3:

	3: class PuppetX::PuppetLabs::Strings::YARD::Handlers::Base < YARD::Handlers::Base

[error]: ArgumentError: superclass #<yardoc class PuppetX::PuppetLabs::Strings::YARD::Handlers::Base> cannot be the same as the declared class #<yardoc class PuppetX::PuppetLabs::Strings::YARD::Handlers::Base>
[error]: Stack trace:
	/Users/iankronquist/.rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/yard-0.8.7.6/lib/yard/code_objects/class_object.rb:139:in `superclass='
	/Users/iankronquist/.rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/yard-0.8.7.6/lib/yard/handlers/ruby/class_handler.rb:18:in `block (2 levels) in <class:ClassHandler>'
	/Users/iankronquist/.rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/yard-0.8.7.6/lib/yard/handlers/base.rb:408:in `block in register'
	/Users/iankronquist/.rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/yard-0.8.7.6/lib/yard/handlers/base.rb:405:in `each'
	/Users/iankronquist/.rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/yard-0.8.7.6/lib/yard/handlers/base.rb:405:in `register'
	/Users/iankronquist/.rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/yard-0.8.7.6/lib/yard/handlers/ruby/class_handler.rb:17:in `block in <class:ClassHandler>'

Installing RDoc documentation for puppetlabs-strings-0.2.0...
iankronquist@puppettop:(strings)(pr/54) → bundle exec irb
irb(main):001:0> require 'strings'
LoadError: cannot load such file -- strings
	from (irb):1:in `require'
	from (irb):1
	from /Users/iankronquist/.rbenv/versions/1.9.3-p551/bin/irb:12:in `<main>'
irb(main):002:0> require 'puppetlabs-strings'
LoadError: cannot load such file -- puppetlabs-strings
	from (irb):2:in `require'
	from (irb):2
	from /Users/iankronquist/.rbenv/versions/1.9.3-p551/bin/irb:12:in `<main>'
irb(main):003:0> require 'puppet-strings'
LoadError: cannot load such file -- puppet-strings
	from (irb):3:in `require'
	from (irb):3
	from /Users/iankronquist/.rbenv/versions/1.9.3-p551/bin/irb:12:in `<main>'
irb(main):004:0> 

```

I see three possible issues here:
1. building the gem warns about the lack of a maintainer email and the description and summary being the same. I will reach out to find the right email. I'm not sure what to do about the description and summary being the same.
2. It totally explodes when generating documentation. That's our fault, it probably deserves a new bug.
3. After I install it I can't require the gem. Strangely enough, if I hack it up so that the gem is named strings I can `require 'strings'`.

iankronquist commented:
I think the email can be `info@puppetlabs.com`.

danzilio commented:
@iankronquist I don't think you'd be able to `require puppetlabs-strings` as there's nothing in this gem at `lib/puppetlabs-strings.rb`. You should, however, be able to `require puppet-strings/rake_tasks` and generate documentation with `PuppetX::PuppetLabs::Strings::Util.generate`. Here's what I'm doing:

```
$ bundle exec gem build puppet-strings.gemspec
WARNING:  no email specified
WARNING:  description and summary are identical
WARNING:  open-ended dependency on puppet (>= 3.7.0) is not recommended
  if puppet is semantically versioned, use:
    add_runtime_dependency 'puppet', '~> 3.7', '>= 3.7.0'
WARNING:  See http://guides.rubygems.org/specification-reference/ for help
  Successfully built RubyGem
  Name: puppetlabs-strings
  Version: 0.2.0
  File: puppetlabs-strings-0.2.0.gem

$ bundle exec gem install puppetlabs-strings-0.2.0.gem
Successfully installed puppetlabs-strings-0.2.0
1 gem installed

$ bundle exec puppet strings ~/git/puppetlabs-apache
Files:           8
Modules:         0 (    0 undocumented)
Classes:         1 (    1 undocumented)
Constants:       1 (    1 undocumented)
Methods:        12 (    5 undocumented)
Puppet Classes:     0 (    0 undocumented)
Puppet Types:     0 (    0 undocumented)
 50.00% documented
```

You don't see the `rdoc` warnings for me because I have `--no-ri --no-rdoc` in my `.gemrc`. I've added this gem to the `Gemfile` of my VirtualBox module (https://github.com/danzilio/danzilio-virtualbox). Here's the commit to add the gem: https://github.com/danzilio/danzilio-virtualbox/commit/526542bd00e0b3c3b8e6cf314f29b4c050bf4bb8. Here's the output of testing with the rake tasks (I dropped my Gemfile.lock for good measure):

```
$ git clone https://github.com/danzilio/danzilio-virtualbox
Cloning into 'danzilio-virtualbox'...
remote: Counting objects: 573, done.
remote: Total 573 (delta 0), reused 0 (delta 0), pack-reused 573
Receiving objects: 100% (573/573), 161.07 KiB | 0 bytes/s, done.
Resolving deltas: 100% (345/345), done.
Checking connectivity... done.

$ cd danzilio-virtualbox/

$ rm Gemfile.lock

$ bundle install
Fetching https://github.com/danzilio/puppetlabs-strings.git
Fetching gem metadata from https://rubygems.org/.........
Fetching additional metadata from https://rubygems.org/..
Resolving dependencies...
Using rake 10.4.2
Using CFPropertyList 2.2.8
Using i18n 0.7.0
Using json 1.8.3
Using minitest 5.8.0
Using thread_safe 0.3.5
Using tzinfo 1.2.2
Using activesupport 4.2.4
Using addressable 2.3.8
Using extlib 0.9.16
Using multi_json 1.11.2
Using autoparse 0.3.3
Using mini_portile 0.6.2
Using nokogiri 1.6.6.2
Using aws-sdk-v1 1.65.0
Using aws-sdk 1.65.0
Using backports 3.6.6
Using require_all 1.3.2
Using stringify-hash 0.0.2
Using beaker-answers 0.0.0
Using excon 0.45.4
Using docker-api 1.22.3
Using fission 0.5.0
Using builder 3.2.2
Using formatador 0.2.5
Using mime-types 2.6.1
Using net-ssh 2.9.2
Using net-scp 1.2.1
Using fog-core 1.32.1
Using fog-xml 0.1.2
Using fog-atmos 0.1.0
Using fog-json 1.0.2
Using ipaddress 0.8.0
Using fog-aws 0.7.5
Using inflecto 0.0.2
Using fog-brightbox 0.9.0
Using fog-dynect 0.0.2
Using fog-ecloud 0.1.1
Using fog-google 0.0.7
Using fog-local 0.2.1
Using fog-powerdns 0.1.1
Using fog-profitbricks 0.0.5
Using fog-radosgw 0.0.4
Using fog-riakcs 0.1.0
Using fog-sakuracloud 1.0.1
Using fog-serverlove 0.1.2
Using fog-softlayer 0.4.7
Using fog-storm_on_demand 0.1.1
Using fog-terremark 0.1.0
Using fog-vmfusion 0.1.0
Using fog-voxel 0.1.0
Using fog 1.33.0
Using multipart-post 2.0.0
Using faraday 0.9.1
Using jwt 1.5.1
Using little-plugger 1.1.3
Using logging 2.0.0
Using memoist 0.12.0
Using signet 0.6.1
Using googleauth 0.4.2
Using launchy 2.4.3
Using retriable 1.4.1
Using google-api-client 0.8.6
Using hocon 0.9.3
Using inifile 2.0.2
Using open_uri_redirections 0.2.1
Using trollop 2.1.2
Using rbvmomi 1.8.2
Using rsync 1.0.9
Using unf_ext 0.0.7.1
Using unf 0.1.4
Using beaker 2.20.0
Using rspec-support 3.1.2
Using rspec-core 3.1.7
Using diff-lcs 1.2.5
Using rspec-expectations 3.1.2
Using rspec-mocks 3.1.3
Using rspec 3.1.0
Using rspec-its 1.2.0
Using net-telnet 0.1.1
Using sfl 2.2
Using specinfra 2.42.1
Using serverspec 2.22.0
Using beaker-rspec 5.2.0
Using coderay 1.1.0
Using domain_name 0.5.24
Using ffi 1.9.10
Using ethon 0.7.4
Using facter 2.4.4
Using faraday_middleware 0.10.0
Using net-http-persistent 2.9.4
Using net-http-pipeline 1.0.1
Using gh 0.14.0
Using rb-fsevent 0.9.5
Using rb-inotify 0.9.5
Using listen 3.0.3
Using lumberjack 1.0.9
Using nenv 0.2.0
Using shellany 0.0.1
Using notiffany 0.0.7
Using method_source 0.8.2
Using slop 3.6.0
Using pry 0.9.12.6
Using thor 0.19.1
Using guard 2.13.0
Using guard-rake 1.0.0
Using json_pure 1.8.2
Using hiera 3.0.1
Using highline 1.7.3
Using http-cookie 1.0.2
Using metaclass 0.0.4
Using spdx-licenses 1.0.0
Using metadata-json-lint 0.0.6
Using mocha 1.1.0
Using netrc 0.10.3
Using puppet 4.2.1
Using rest-client 1.8.0
Using puppet-blacksmith 3.3.1
Using puppet-lint 1.1.0
Using puppet-syntax 2.0.0
Using yard 0.8.7.6
Using puppetlabs-strings 0.2.0 from https://github.com/danzilio/puppetlabs-strings.git (at gem)
Using rspec-puppet 2.2.0
Using puppetlabs_spec_helper 0.10.3
Using websocket 1.2.2
Using pusher-client 0.6.2
Using typhoeus 0.7.3
Using travis 1.8.0
Using travis-lint 2.0.0
Using vagrant-wrapper 2.0.3
Using bundler 1.7.2
Your bundle is complete!
Use `bundle show [gemname]` to see where a bundled gem is installed.

$ bundle exec rake -T
rake acceptance                              # Run acceptance tests
rake beaker                                  # Run beaker acceptance tests
rake beaker_nodes                            # List available beaker nodesets
rake build                                   # Build puppet module package
rake clean                                   # Clean a built module package
rake coverage                                # Generate code coverage information
rake help                                    # Display the list of available rake tasks
rake lint                                    # Run puppet-lint
rake metadata                                # Validate metadata.json file / Run metadata-json-lint
rake module:bump                             # Bump module version to the next patch
rake module:bump:major                       # Bump module version to the next MAJOR version
rake module:bump:minor                       # Bump module version to the next MINOR version
rake module:bump:patch                       # Bump module version to the next PATCH version
rake module:bump_commit                      # Bump version and git commit
rake module:clean                            # Runs clean again
rake module:dependency[module_name,version]  # Set specific module dependency version
rake module:push                             # Push module to the Puppet Forge
rake module:release                          # Release the Puppet module, doing a clean, build, tag, push, bump_commi...
rake module:tag                              # Git tag with the current module version
rake spec                                    # Run spec tests in a clean fixtures directory
rake spec_clean                              # Clean up the fixtures directory
rake spec_prep                               # Create the fixtures directory
rake spec_standalone                         # Run spec tests on an existing fixtures directory
rake strings:generate                        # Generate Puppet documentation with YARD
rake strings:serve                           # Serve YARD documentation for modules
rake syntax                                  # Syntax check Puppet manifests and templates
rake syntax:hiera                            # Syntax check Hiera config files
rake syntax:manifests                        # Syntax check Puppet manifests
rake syntax:templates                        # Syntax check Puppet templates
rake test                                    # Run syntax, lint, and spec tests
rake validate                                # Check syntax of Ruby files and call :syntax and :metadata

$ bundle exec rake strings:generate
Files:           5
Modules:         0 (    0 undocumented)
Classes:         0 (    0 undocumented)
Constants:       0 (    0 undocumented)
Methods:         0 (    0 undocumented)
Puppet Classes:     4 (    0 undocumented)
Puppet Types:     1 (    0 undocumented)
 100.00% documented
```

danzilio commented:
You should also be able to `require 'puppet_x/puppetlabs/strings/util'`

iankronquist commented:
Okay, that make sense. I'd say add the maintainer email and you'll have my +1. However, I'd like to get another set of eyes on this and others are either out of town on vacation, or up to their nose in work. I'll try to get someone else to review this when they can.

danzilio commented:
Great! I'll push that now!

HAIL9000 commented:
Hey @danzilio, thanks so much for working on this! It looks like (I'm guessing due to README changes) this is in need of a rebase. We also like to make sure that all of our commits reference a ticket, so would you mind updating the commit messages to do this? We have a ticket filed for this work [PDOC-28](https://tickets.puppetlabs.com/browse/PDOC-28). There should be a good description of how to do that in our [contributing guidelines](https://github.com/puppetlabs/puppetlabs-strings/blob/master/CONTRIBUTING.md).

Thanks again! c:

HAIL9000 commented:
@iankronquist Now that this is updated could you take another look and see what you think?

iankronquist commented:
Looks good to me. I think we will want to change the Yard version to 0.8.7.6, the current version of Yard, but that's another discussion.

iankronquist commented:
Actually, if we make this change, and I understand how gem dependencies work, then we should remove the section in the README telling people to install Yard manually.

However, I'm in favor of just merging this and making the README change a separate PR fixing [PDOC-50](https://tickets.puppetlabs.com/browse/PDOC-50).

danzilio commented:
Some users will probably still install strings as a module, unless we want the gem to be the only way to install strings, I think we should leave the README as is. Also, I'm not entirely sure we should pin the YARD version to that specific version. The pessimistic version matcher should be good enough, but I'm not sure how volatile YARD is. Thoughts? I just want to avoid dependency hell.

danzilio commented:
How do we feel about the name of this gem? I'm starting to regret changing it to puppetlabs-strings. Our attempt at consistency may have resulted in a lack of clarity. I think most people refer to it as 'puppet strings' instead of 'puppetlabs strings' but I'd like to hear from some others about this.

HAIL9000 commented:
My feeling is that puppet-strings is likely a better name and more consistent with things like puppet-lint.

ferventcoder commented:
puppet-strings!

danzilio commented:
Ok! While I'm all for internal consistency, unless @iankronquist has strong feelings about it, I'm going to call this a quorum and change it back to puppet-strings.

iankronquist commented:
Go ahead and change it. It's starting to look like the better decision.

danzilio commented:
The motion to commit is adopted! This gem is hereby dubbed 'puppet-strings'

HAIL9000 commented:
Hey @danzilio, sorry this fell off the map a little bit. I'll pick this back up next week after PuppetConf so we can get this merged. :)

danzilio commented:
No worries @HAIL9000 :) See you at PuppetConf!

HAIL9000 commented:
...Actually this looks good to me so I'm going to go ahead and merge it so we can keep things moving on the gem release
